Handle dependencies

anotherjames opened the pull request:
Magic!
- Rename the keys we use in the 'manifest' (info file) (and update related code accordingly - see https://www.drupal.org/node/2806197 although that is outside the scope of the MVP)
- Suggest dependants to export, with ability to export them at the same time (by explicitly adding them to the manifest for now)
- Always export dependencies of whatever is being explicitly exported

stevetweeddale committed:
Refactor keys away from devel_config into cm_config_tools

Old keys were
- devel_config
- cm_tools_config.create_only
- cm_tools_config.delete
New keys are:
- cm_tools_config.managed
- cm_tools_config.unmanaged
- cm_tools_config.delete

anotherjames committed:
Fixed merge conflict

anotherjames commented on drush/cm_config_tools.drush.inc:
```diff
@@ -19,12 +19,11 @@ function cm_config_tools_drush_command() {
   $description .= "Write back configuration to module's config/install directory." . "\n";
   $description .= "List which configuration settings you want to export in the" . "\n";
-  $description .= "module's info file by listing them under 'config_devel', as shown below:"  . "\n";
+  $description .= "modules info file by listing them under 'cm_config_tools', as shown below:"  . "\n";
```
should be "module's" rather than "modules", sorry! :-)

anotherjames commented:
Unrelated thought btw ... Unless it's not too difficult to implement/maintain, I wonder if we should just drop support for config across multiple modules/buckets?

If it can be done, it would be nice, but I can see it getting tricky when it comes to exporting dependencies, as if module A and module B both have a particular dependency, how do we handle that? Does the config get exported to both, do we arbitrarily pick one?
(And the same probably applies to dependants too.)

A 'solution' _could_ be to only support exporting dependencies/dependants when working with just the one module. But that is more or less what we've already got - so I wonder if we should just consider all the work in this PR as a cm_config_tools 'version 2.0' (i.e. 8.x-2.x). That would allow any projects that need that possibility could remain on 1.0, without nice dependency management. And then in future, if we can think of a better solution, that can be added on top of either 'version', or as a 3.0 if it can't really be backwards-compatible.

@stevetweeddale and @darthsteven what do you think of that as a plan?
I'm not raising this to unquestionably encourage us to support splitting config across modules, but it would still be 'nice' to be able to if solutions can be found, even just in future. Or at least just to try & avoid making decisions that break that use case until necessary. Otherwise we may end up pushing in one direction before we're actually 100% sure about it (as we won't have used any of this in practise for a while).

stevetweeddale commented:
@anotherjames presumably as it stands, because we don't have any kind of features-like 'conflict' management, if you try to export the same thing to 2 different modules which ever module comes last will 'win' on import? 

Auto-added dependencies aside, Is there anything to stop me manually adding the same thing to 2 modules with the current version?

stevetweeddale committed:
Fixup outdated interface

stevetweeddale committed:
grammar++

anotherjames committed:
Merge branch '8.x-1.x' into feature/dependencies

stevetweeddale committed:
Fix references to cm_tools_config (should be cm_config_tools)

anotherjames committed:
Fixed merge conflict

stevetweeddale committed:
Add methods to service for retrieving dependancies and dependants

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+   * @return array
+   *   An array of config the extension depends on.
+   */
+  public function getConfigDependancies($extension);
```
spelling in a few places: dependancies => dependencies, sorry!

stevetweeddale replied:
Okidoke - I think either spelling is actually valid, but happy to standardise on "dependencies" as the plural adjective and "dependant" as the singular noun.

anotherjames replied:
loving the irony of the initially-incorrectly-spelt word :-) I think we're talking the same language now though 👍

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+   *
+   * For the config listed in a projects .info.yml, find other config that is
+   * dependant upon it, but which is not:
+   *  - Itself a dependency of the config listed in cm_config_tools.export
```
by 'cm_config_tools.export', do you mean 'cm_config_tools.managed', in which case can this and the following lines be changed to reflect that please?

stevetweeddale replied:
Yup, this was written before we refactored/decided on names. Now fixed.

anotherjames replied:
ah loooovely :-)

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+  public function getDependentConfigSuggestions($extension);
+
+  /**
+   * Adds the provided config keys to an extensions explicitly exported config.
```
grammar: extension's
... and let's say 'explicitly listed' rather than 'explicitly exported' please

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+  /**
+   * Adds the provided config keys to an extensions explicitly exported config.
+   *
+   * Add the config keys to an extensions .info.yml, under
```
grammar: extension's

... also, should the key be 'cm_config_tools.managed' rather than 'cm_config_tools.export'?

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+   * @param $config_keys
+   * @return mixed
+   */
+  public function addConfigKeysToManifest($extension, $config_keys);
```
We haven't referred to the .info.yml file as a manifest anywhere else in code (yet). I suggest that whilst it is true that it's conceptually a manifest, I'm not sure introducing that idea when we're not doing so elsewhere is a good idea. Either we should update various other bits of commenting/code/documentation to refer to it that way, or can we find a better name for this method please (which I think I'd prefer to do)?

In other news, 'config keys' isn't really the way they're referred to elsewhere either -- can we just say 'config names' please?

Sorry to always be picky about naming! I'm aiming for this module to be as intuitive for any CM dev as possible, hence documentation & naming & consistency being important.

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
     return $source_storage;
   }
 
+  public function sortAndFilterOutput($dependencies) {
```
does this need to be a public method? I don't think it does (but if it did, it should probably be in the interface too).

anotherjames commented on src/ExtensionConfigHandlerInterface.php:
```diff
@@ -107,6 +108,47 @@ public function getStorageComparer(array $source_dirs, $subdir = InstallStorage:
+   * @return array
+   *   An array of config that depends on config currently exported.
+   */
+  public function getDependentConfigSuggestions($extension);
```
Please can we rename this either to getSuggestedConfig (i.e. this function will suggest config, and the implementing class will decide what that means) or getDependants() (i.e. the function will get dependants and then calling code can decide what to do with that), or something else similar? The current function name feels like it hints too much at implementation as well as usage, I think it should be one or the other. 
(And having 'Dependent' in the name causes more confusion as I think we refer to them as depend**a**nts everywhere else where possible, i.e. with the 'a', to distinguish them from dependencies)

anotherjames replied:
Note that in another comment I also suggest that we rename this to be clear that its dealing with a whole extension, so in that case, I suggest calling this getExtensionDependants() (or getExtensionSuggestedConfig / getSuggestedExtensionConfig )?

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfigDependancies($extension, $limit = NULL) {
```
$limit parameter is not in the interface...

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+   */
+  public function getConfigDependancies($extension, $limit = NULL) {
+    $dependencies = array();
+    $listed_config = $this->getExtensionInfo($extension, 'managed');
```
getExtensionInfo() can't be assumed to return an array (e.g. if the extension is not enabled/found, or if invalid stuff is in the info file. At least add `if ($listed_config && is_array($listed_config))` before foreach-ing, and probably also add `array()` as a 3rd $default parameter to getExtensionInfo() - thanks!

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+  }
+
+  /**
+   * @param $name
```
Function doxygen needed please :-)

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+   *   The config key to find the dependencies of.
+   * @param null $recursion_limit
+   *   Optionally limit the levels of recursion.
+   * @return mixed
```
Please can we flesh out the @return doxygen here?

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+   *   Optionally limit the levels of recursion.
+   * @return mixed
+   */
+  public function getIndividualConfigDependancies($name, $recursion_limit = NULL) {
```
I wonder if it's worth renaming this and the getConfigDependancies/getConfigDependencies function to be clearer. This one gets dependencies for a specific item of config, the other gets dependencies for all the config in an extension. So how about changing the other to be getExtensionConfigDependencies() and then this one is actually truer to be called getConfigDependencies ?
(...and then update the interface appropriately too, so that the one that covers the whole extension is the one in the interface)

More importantly, please can we make this function protected, or if there's a good reason not to (?), add it to the interface too?

stevetweeddale replied:
I left it public because I thought it might be helpful to have accessible. What's achieved by protecting it?

On the other hand I didn't think it necessary to require it of _any_ implementor of our interface, so had left it out. But if you'd like things either protected or in the interface, I'll go for the interface.

anotherjames replied:
Ok sure - let's add it to the interface then too please.

I guess I had felt like it didn't _really_ do much beyond what core provides easily already, but it's the recursive checking of dependencies that is actually important to us, and is therefore useful enough to be made public. ...and as I'd consider any other implementor to need to implement what we consider useful, I'd put it in the interface - although yeah, not strictly necessary, agreed. 

I recognise I'm being a bit backwards in deciding what should be in the interface by what's in our implementation, but that's all a bit arbitrary anyway as it's a custom service that's unlikely to ever get other implementors. The interface then just becomes a useful place to 'define' our service without being mixed with the actual implementation code.

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+        $config_dependencies = array();
+      }
+
+      // Config provider is an implied module dependency.
```
I don't know how easy it is to establish, but there's a chance that the config could be from a theme! Maybe a call to getExtensionType is needed? (If that returned 'profile' or anything other than 'theme', then it would be fine to use 'module' here)

stevetweeddale replied:
Left this one for you

anotherjames replied:
👍  :-)

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+  public function getDependentConfigSuggestions($extension, $recursion_limit = NULL) {
+    $dependants = array();
+    $listed_config = $this->getExtensionInfo($extension, 'managed');
+    $dependency_manager = \Drupal::service('config.manager')
```
Please can we use `$this->configManager` instead of `\Drupal::service('config.manager') here? Ta!

stevetweeddale replied:
Also, DIY :)

anotherjames replied:
👍  :-)

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,156 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+   * @param null $recursion_limit
+   * @return mixed
+   */
+  public function getIndividualConfigDependants($key, $dependency_manager, $recursion_limit = NULL) {
```
As mentioned for the equivalent functions for dependencies, could we rename this and the getDependentConfigSuggestions functions to be clearer that one deals with an extension as a whole, and only the other is for a config item? There was a comment around renaming getDependentConfigSuggestions() too, so let's figure that one out first, but that one should really have `ExtensionConfig` in the name I think, whereas its this one that could just be `somethingConfigSomething`.
...and the function should be protected (+1 from me) or in the interface (I could be persuaded of this).

anotherjames reviewed (commented):
I realise this is all work-in-progress, so I'm sorry to jump in on stuff you're probably already aware of. Thanks for the great work -- sorry for the many nitpicks!

stevetweeddale committed:
Fix typo

stevetweeddale committed:
Standardise on 'depencencies' (adjective) and 'dependants' (noun)

stevetweeddale committed:
Update doxygen to reflect changed keys

stevetweeddale committed:
Rename methods.

1)
getConfigDependencies() is now:
getExtensionConfigDependencies()

2)
getDependentConfigSuggestions() is now:
getExtensionConfigSuggestions()

… which nicely matches it's 'inverse' partner, 1)

stevetweeddale committed:
Protect sortAndFilterOutput()

stevetweeddale committed:
Rename unimplemented placeholder method, feel free to change

stevetweeddale committed:
Remove 'individual' from sub methods

anotherjames commented on src/ExtensionConfigHandler.php:
```diff
@@ -352,6 +352,167 @@ protected function getSourceStorageWrapper(array $source_dirs, $subdir = Install
+          $recursive_iterations++;
+          if ($recursion_limit && $recursive_iterations < $recursion_limit) {
+            foreach ($config_dependencies['config'] as $dependency) {
+              $sub_dependencies = $this->getIndividualConfigDependencies($dependency, $recursion_limit);
```
This needs to be `getConfigDependencies()` now?

stevetweeddale committed:
Remove legacy 'individual' from method call

anotherjames committed:
Merge branch '8.x-1.x' into feature/dependencies

anotherjames committed:
Lots of change, for version 2.0.

* UNTESTED!
* Implement suggested changes from 
https://github.com/computerminds/cm_config_tools/pull/1#pullrequestreview-1771930
* Put all public methods into interface.
* Removed ability to work on multiple modules, with the assumption that
  there would only be a single bucket of config, yet potentially looping
  over all extensions. That means unsupported behaviour if there
  actually are multiple extensions using cm_config_tools.
* Updated parameters & documentation, for consistency/clarification
* Inject all possible dependencies (config manager/factory and state)

anotherjames committed:
Added more TODO notes.

anotherjames committed:
Exclude listed config from suggestions.

anotherjames commented:
Time to merge :-) Nothing is broken AFAIK, so next steps can just happen in normal dev branch.
